OpenRA.Mods.Common/Traits/BotModules/PowerDownBotManager.cs

```diff
@@ -0,0 +1,201 @@
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[TraitLocation(SystemActors.Player)]
+	[Desc("Manages AI powerdown.",
+		"You need to use PowerMultiplier on toggle control only on related buildings, for calculation of this bot module")]
+	public class PowerDownBotModuleInfo : ConditionalTraitInfo
+	{
+		[Desc("Delay (in ticks) between two action on toggling powerdown.")]
+		public readonly int Interval = 150;
+
+		[Desc("Actors that allow this module to toggle")]
+		public readonly HashSet<string> PowerDownTypes = [];
+
+		[Desc("Order used by " + nameof(ToggleConditionOnOrderInfo) + " for powerdown on toggled actor.")]
+		public readonly string PowerDownOrder = "PowerDown";
+
+		public override object Create(ActorInitializer init) { return new PowerDownBotModule(init.Self, this); }
+	}
+
+	public class PowerDownBotModule : ConditionalTrait<PowerDownBotModuleInfo>, IBotTick, IGameSaveTraitData, INotifyActorDisposing
+	{
+		readonly World world;
+		readonly Player player;
+		readonly ActorIndex.OwnerAndNamesAndTrait<ToggleConditionOnOrderInfo> togglable;
+		readonly Func<Actor, bool> isTogglableValid;
+
+		PowerManager playerPower;
+		int toggleTick;
+
+		// We keep a list to track toggled buildings for performance.
+		List<ToggledPowerWrapper> toggled = [];
+
+		sealed class ToggledPowerWrapper
+		{
+			public int ExpectedPowerChanging;
+			public Actor Actor;
+
+			public ToggledPowerWrapper(Actor a, int p)
+			{
+				Actor = a;
+				ExpectedPowerChanging = p;
+			}
+		}
+
+		public PowerDownBotModule(Actor self, PowerDownBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+			togglable = new ActorIndex.OwnerAndNamesAndTrait<ToggleConditionOnOrderInfo>(world, info.PowerDownTypes, player);
+
+			isTogglableValid = a => a != null && a.Owner == self.Owner && !a.IsDead && a.IsInWorld;
+		}
+
+		protected override void Created(Actor self)
+		{
+			playerPower = self.Owner.PlayerActor.TraitOrDefault<PowerManager>();
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(Info.Interval);
+		}
+
+		// We calculate the approximate power changing if toggled on. Since we have no idea on the conditions set by users
+		// Calculation here need user to use PowerMultiplier on toggle control only on related buildings, for calculation of this bot module.
+		static int GetTogglePowerChanging(Actor a)
+		{
+			var powerChangingIfToggled = 0;
+			var powerTraits = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).ToArray();
+			if (powerTraits.Length > 0)
```

Comment:
```suggestion
			var power = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).Sum(t => t.Info.Amount);
			if (power != 0)
```

perhaps do this for less memory allocation?

Reply:
Fixed

Reply:
good point